postprocessing/resstockpostproc/process_metadata.py

```diff
@@ -9,62 +9,96 @@ def get_failed_bldgs(metadata_df: pl.LazyFrame) -> set[int]:
+    upgrade_df = upgrade_raw_df.filter((~pl.col("building_id").is_in(baseline_failed_bldgs)) &
                               (pl.col("completed_status") == "Success"))
+    failed_bldgs = (
+        upgrade_raw_df.filter(
+            (pl.col("completed_status") == "Fail")
+            & (~pl.col("building_id").is_in(baseline_failed_bldgs))
+        )
+        .select(pl.col("building_id"))
+        .collect()['building_id']
+        .to_list()
+    )
+    if failed_bldgs:
+        print(
+            f"Replacig these {len(failed_bldgs)} buildings that only failed in "
+            f"upgrade {upgrade_num} with baseline: {failed_bldgs}"
+        )
     
-    upgrade = get_transformed_cols(upgrade, col_maps)
-    upgrade = upgrade.with_columns([pl.lit(upgrade_num).alias("upgrade")])
-    base_cols = base.collect_schema().names()
-    upgrade_cols = upgrade.collect_schema().names()
+    upgrade_df = get_transformed_cols(upgrade_df, col_maps)
+    upgrade_df = upgrade_df.with_columns([pl.lit(upgrade_num).alias("upgrade")])
+    base_cols = base_pub_df.collect_schema().names()
+    upgrade_cols = upgrade_df.collect_schema().names()
     missing_cols = list(set(base_cols) - set(upgrade_cols)) + ["bldg_id"]
-    upgrade = upgrade.join(base.select(missing_cols), on="bldg_id", how="left")
-    all_cols = upgrade.collect_schema().names()
+    upgrade_df = upgrade_df.join(base_pub_df.select(missing_cols), on="bldg_id", how="left")
+    all_cols = upgrade_df.collect_schema().names()
     print("Fixing site energy and site emission total for upgrade ...")
-    upgrade = fix_site_energy_total(upgrade, all_cols)
-    upgrade = fix_all_fuels_emissions(upgrade, all_cols)
-    upgrade = add_upgrade_columns(upgrade)
-    upgrade = upgrade.with_columns(pl.lit(True).alias("applicability"))
+    upgrade_df = fix_site_energy_total(upgrade_df, all_cols)
+    upgrade_df = fix_all_fuels_emissions(upgrade_df, all_cols)
+    upgrade_df = add_upgrade_columns(upgrade_df)
+    upgrade_df = upgrade_df.with_columns(pl.lit(True).alias("applicability"))
     # get upgrade_name
-    upgrade_name_df = upgrade.select(pl.col("upgrade_name").first())
-    missing_bldgs_df = base.join(
-        upgrade,
+    upgrade_name_df = upgrade_df.select(pl.col("upgrade_name").first())
+    missing_bldgs_df = base_pub_df.join(
+        upgrade_df,
         on="bldg_id",
         how="anti"  # Keep rows from 'base' with no match in 'upgrade'
     )
     missing_bldgs_df = missing_bldgs_df.with_columns([
         pl.lit(False).alias("applicability"),
         pl.lit(upgrade_num).alias("upgrade"),
     ]).drop("upgrade_name")
-    upgrade_cols = upgrade.collect_schema().names()
+    upgrade_cols = upgrade_df.collect_schema().names()
     missing_bldgs_df = missing_bldgs_df.join(upgrade_name_df,  how="cross")
-    upgrade = pl.concat([upgrade, missing_bldgs_df], how="diagonal_relaxed")
-    upgrade = upgrade.sort("bldg_id")
-    upgrade = add_saving_cols(upgrade, base)
```

Comment:
I think this needs to be called on the baseline DF as well so that that schemas for baseline and upgrade files are the same. Even though the savings will be zero, this is easier for users.